### PR TITLE
[WIP] Avinetworks unit testing using vcr

### PR DESCRIPTION
**Pull request by grastogi23**
##### SUMMARY

Acceptance Tests for Avi Modules

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

network/avi

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = /home/grastogi/.ansible.cfg
  configured module search path = [u'/home/grastogi/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.6 (default, Jun 22 2015, 17:58:13) [GCC 4.8.2]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by grastogi23** [comment]
Hi @gundalow @mattclay,
As discussed before, pl. do not merge this yet!. We can go over this in next Ansible Network meeting. 

thanks

File: test/units/modules/network/avi/avi_healthmonitor_test.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import json
+import vcr
```
Comment (gundalow) [inline test/units/modules/network/avi/avi_healthmonitor_test.py]: We will want to skip test if this import fails. I'll find an example you can copy.

This makes it cleaner for people running the unit tests locally without `requirements/unit.txt` installed

Comment (gundalow) [reply test/units/modules/network/avi/avi_healthmonitor_test.py]:
```python
try:
    import vcr
except ImportError:
    raise SkipTest("vcr is needed")
```

**Review by mattclay — CHANGES_REQUESTED** [review]
Given that the AVI modules and module_utils are extremely thin wrappers around the AVI SDK, I don't think we're getting much value out of replaying VCR based tests in this case.

Looking at the modules and the code coverage we're already getting from import tests, I think there are two kinds of tests that would be beneficial:

1. Integration Tests - Standard integration test playbooks would test the integration between the modules and the SDK. The same thing accomplished by recording the VCR tests.
2. Unit Tests - Unit tests using mock for the avisdk to permit code coverage of branches not easily tested by the integration tests.

**Comment by grastogi23** [comment]
@mattclay 
Actually, VCR tests are indeed the integration tests except that there are no real Avi Controller behind it but a recording of it. The recording in VCR is created from a real Avi Controller endpoint. So, it is as though we have frozen Avi Controller and bundled in these tests. This way both tests run fast and don't have to do complex integration. 

As far as the additional unit test is concerned, there is not much we will get if we mock out the avisdk as everything comes down to single api avi_ansible_api. 

We will try to do live discussion next wednesday to better explain role of vcr and how it is indeed integration test.

**Comment by mattclay** [comment]
@grastogi23 My point was that the creation of the VCR cassette is the actual test. Re-running the tests by replaying a cassette doesn't provide any additional testing.

**Comment by bdowling** [comment]
As mentioned, the point of vcrpy is to record the web API interaction and then automatically simulate the endpoint in the future for must faster regression testing.  You are testing (without needing the DuT and obvious speed improvements) that your contract with the DuT API and handling it's output remains intact after that point.

I haven't looked at this PR in detail, but I like the idea of automatically mocking out the APIs from the actual devices.  Anytime the device API changes you can remove the cassettes and regenerate.  An integration test that validates cassettes against the live API is also useful in cases where the API is changing.

I have considered that a similar auto-mock capability could/should exist for any other iostreams, such as CLI for example.  Just my $.02 -- I had asked about the vcrpy meeting topic in #ansible-network  the other day and was pointed at this as one PR potentially using that..

**Comment by mattclay** [comment]
I understand that VCR may provide value for some test scenarios. However, I don't believe this is one of them. Since there is almost no logic in the AVI modules, there isn't anything to test that wasn't already tested during the VCR recording. Nearly any change made to an AVI module will result in invalidating the VCR cassettes and would thus require creating a new VCR cassette.

**Comment by grastogi23** [comment]
@mattclay @bdowling Also, current PR has very simple test example. We will add a more realistic scenario as that may also help in discussion. we will discuss it on wednesday meeting. 

ps: @chaitanyaavi

File: requirements.txt [hunk 2]
```diff
@@ -8,3 +8,4 @@ PyYAML
 paramiko
 cryptography
 setuptools
+vcrpy
```
Comment (mattclay) [inline requirements.txt:11]: These are requirements for ansible, not for tests. No new requirements should be added here.

**Pull request closed without merging**